src/Cache.js

```diff
@@ -135,16 +135,20 @@ define([
 			});
 		},
 		put: function (object, directives) {
-			// first remove from the cache, so it is empty until we get a response from the master store
+			// first update the cache, we are going to assume the update is valid while we wait to
+			// hear from the master store
 			var cachingStore = this.cachingStore;
-			cachingStore.remove((directives && directives.id) || this.getIdentity(object));
-			return when(this.inherited(arguments), function (result) {
+			cachingStore.put(object, directives);
+			return when(this.inherited(arguments)).then(function (result) {
 				// now put result in cache
 				var cachedPutResult =
 					cachingStore.put(object && typeof result === 'object' ? result : object, directives);
```

Comment:
This can be reduced to an `if`, since if result is not an object, we've already done this prior to calling inherited.